pkg/retainer/txretainer/tx_retainer.go

```diff
@@ -0,0 +1,387 @@
+		return false
+	}
+
+	if m.EarliestAttachmentSlot != other.EarliestAttachmentSlot {
+		return false
+	}
+
+	if m.State != other.State {
+		return false
+	}
+
+	if m.FailureReason != other.FailureReason {
+		return false
+	}
+
+	if (m.ErrorMsg == nil || other.ErrorMsg == nil) && m.ErrorMsg != other.ErrorMsg {
+		return false
+	}
+
+	if (m.ErrorMsg != nil && other.ErrorMsg != nil) && *m.ErrorMsg != *other.ErrorMsg {
+		return false
+	}
+
+	return true
+}
+
+func (m *TransactionMetadata) String() string {
+	return fmt.Sprintf("tx metadata => TxID: %s, CreatedAtSlot: %d, ValidSignature: %t, State: %d, Error: %d, ErrorMsg: \"%s\"", hex.EncodeToString(m.TransactionID), m.EarliestAttachmentSlot, m.ValidSignature, m.State, m.FailureReason, *m.ErrorMsg)
+}
+
+type (
+	SlotFunc             func() iotago.SlotIndex
+	txMetadataUpdateFunc func(oldTxMeta *TransactionMetadata, newTxMeta *TransactionMetadata) (*TransactionMetadata, bool, error)
+)
+
+// TransactionRetainer keeps and resolves all the transaction-related metadata needed in the API and INX.
+type TransactionRetainer struct {
+	workerPool              *workerpool.WorkerPool
+	txRetainerDatabase      *transactionRetainerDatabase
+	latestCommittedSlotFunc SlotFunc
+	finalizedSlotFunc       SlotFunc
+	errorHandler            func(error)
+
+	txRetainerCache *transactionRetainerCache
+
+	storeDebugErrorMessages bool
+
+	module.Module
+}
+
+// WithDebugStoreErrorMessages configures the TransactionRetainer to store debug error messages.
+func WithDebugStoreErrorMessages(store bool) options.Option[TransactionRetainer] {
+	return func(r *TransactionRetainer) {
+		r.storeDebugErrorMessages = store
+	}
+}
+
+func New(workersGroup *workerpool.Group, dbExecFunc storage.SQLDatabaseExecFunc, latestCommittedSlotFunc SlotFunc, finalizedSlotFunc SlotFunc, errorHandler func(error), opts ...options.Option[TransactionRetainer]) *TransactionRetainer {
+	return options.Apply(&TransactionRetainer{
+		workerPool:              workersGroup.CreatePool("TxRetainer", workerpool.WithWorkerCount(1)),
+		txRetainerCache:         NewTransactionRetainerCache(),
+		txRetainerDatabase:      NewTransactionRetainerDB(dbExecFunc),
+		latestCommittedSlotFunc: latestCommittedSlotFunc,
+		finalizedSlotFunc:       finalizedSlotFunc,
+		errorHandler:            errorHandler,
+	}, opts,
+		(*TransactionRetainer).TriggerConstructed,
+		(*TransactionRetainer).TriggerInitialized,
+	)
+}
+
+// NewProvider creates a new TransactionRetainer provider.
+func NewProvider(opts ...options.Option[TransactionRetainer]) module.Provider[*engine.Engine, retainer.TransactionRetainer] {
+	return module.Provide(func(e *engine.Engine) retainer.TransactionRetainer {
+		r := New(e.Workers.CreateGroup("TransactionRetainer"),
+			e.Storage.TransactionRetainerDatabaseExecFunc(),
+			func() iotago.SlotIndex {
+				// use settings in case SyncManager is not constructed yet.
+				if e.SyncManager == nil {
+					return e.Storage.Settings().LatestCommitment().Slot()
```

Comment:
why is this needed? just change the order in the engine providing part

Reply:
I stole that part from the block retainer :D
I changed the order now and removed it in both retainers :+1: